### ISSUE TITLE
loss odd

### ISSUE THREAD

**Issue by DUZHUJUN** [zylo117/Yet-Another-EfficientDet-Pytorch#40]
Hi,Thanks for your great job first,and I have a loss problem,wish you can help me:
I used my custom dataset,and I have verified its' Correctness，but when I train the model,the loss is very odd, What's wrong with my operation? （the json file i wrote to yml file ）

`python train.py -c 7 -n 2 --batch_size 1 --lr 0.00025 --optim sgd --es_
patience 10 --data_path '/' --log_path /efficientdet/logs/ --load_weights /efficientdet/weights/efficientdet-d7.pth --saved_path /results/efficientdet/results/ -p 'dla'
`

loss：It's either zero or several values are repeated,especially Cls_loss
`Step: 31. Epoch: 0/500. Iteration: 39/400. Cls loss: 2.30212. Reg loss: 0.64907. Total loss: 2.95119:  10%|███▏                              | 38/400 [00:30<04:59,  1.21it/s
Step: 31. Epoch: 0/500. Iteration: 39/400. Cls loss: 2.30212. Reg loss: 0.64907. Total loss: 2.95119:  10%|███▎                              | 39/400 [00:30<05:06,  1.18it/s
Step: 32. Epoch: 0/500. Iteration: 40/400. Cls loss: 2.30212. Reg loss: 1.08668. Total loss: 3.38880:  10%|███▎                              | 39/400 [00:31<05:06,  1.18it/s
Step: 32. Epoch: 0/500. Iteration: 40/400. Cls loss: 2.30212. Reg loss: 1.08668. Total loss: 3.38880:  10%|███▍                              | 40/400 [00:31<05:11,  1.16it/s
Step: 33. Epoch: 0/500. Iteration: 41/400. Cls loss: 0.00000. Reg loss: 0.00000. Total loss: 0.00000:  10%|███▍                              | 40/400 [00:32<05:11,  1.16it/s
Step: 33. Epoch: 0/500. Iteration: 41/400. Cls loss: 0.00000. Reg loss: 0.00000. Total loss: 0.00000:  10%|███▍                              | 41/400 [00:32<05:09,  1.16it/s
Step: 34. Epoch: 0/500. Iteration: 42/400. Cls loss: 2.30212. Reg loss: 1.13490. Total loss: 3.43702:  10%|███▍                              | 41/400 [00:33<05:09,  1.16it/s
Step: 35. Epoch: 0/500. Iteration: 43/400. Cls loss: 0.00000. Reg loss: 0.00000. Total loss: 0.00000:  11%|███▋                              | 43/400 [00:34<05:09,  1.15it/s
Step: 36. Epoch: 0/500. Iteration: 44/400. Cls loss: 0.00000. Reg loss: 0.00000. Total loss: 0.00000:  11%|███▋                              | 43/400 [00:35<05:09,  1.15it/s
Step: 36. Epoch: 0/500. Iteration: 44/400. Cls loss: 0.00000. Reg loss: 0.00000. Total loss: 0.00000:  11%|███▋                              | 44/400 [00:35<05:07,  1.16it/s
Step: 37. Epoch: 0/500. Iteration: 45/400. Cls loss: 2.30212. Reg loss: 1.03828. Total loss: 3.34040:  11%|███▋                              | 44/400 [00:35<05:07,  1.16it/s
Step: 37. Epoch: 0/500. Iteration: 45/400. Cls loss: 2.30212. Reg loss: 1.03828. Total loss: 3.34040:  11%|███▊                              | 45/400 [00:35<05:10,  1.15it/s
Step: 38. Epoch: 0/500. Iteration: 46/400. Cls loss: 2.30212. Reg loss: 0.84100. Total loss: 3.14313:  11%|███▊                              | 45/400 [00:36<05:10,  1.15it/s
Step: 38. Epoch: 0/500. Iteration: 46/400. Cls loss: 2.30212. Reg loss: 0.84100. Total loss: 3.14313:  12%|███▉                              | 46/400 [00:36<05:11,  1.14it/s
Step: 39. Epoch: 0/500. Iteration: 47/400. Cls loss: 2.30212. Reg loss: 0.90405. Total loss: 3.20618:  12%|███▉                              | 46/400 [00:37<05:11,  1.14it/s
Step: 39. Epoch: 0/500. Iteration: 47/400. Cls loss: 2.30212. Reg loss: 0.90405. Total loss: 3.20618:  12%|███▉                              | 47/400 [00:37<05:12,  1.13it/s
Step: 39. Epoch: 0/500. Iteration: 47/400. Cls loss: 2.30212. Reg loss: 0.90405. Total loss: 3.20618:  12%|███▉                              | 47/400 [00:37<04:44,  1.24it/s
]
`

**Comment by zylo117** [comment]
your batchsize is too small, so BN will be very unstable, you should freeze BN.
but freezing bn, the training process will also be unstable.
The only way to solve it, is to utilize multiple gpus, so that syncBN can kick in.
If you don't have multiple gpus, I suggest you switch to smaller network like d2/d3, which enables you to use bigger batchsize

**Comment by ChrisHaooooooo** [comment]
@zylo117 Perhaps a single argument --accumulate could be set to solve the problem？We do not carry out back propagation every batch_size, and then accumulate the loss for --accumulate times , which is equivalent to a large batch_size. I saw the method in [https://github.com/ultralytics/yolov3](url)

**Comment by zylo117** [comment]
> @zylo117 Perhaps a single argument --accumulate could be set to solve the problem？We do not carry out back propagation every batch_size, and then accumulate the loss for --accumulate times , which is equivalent to a large batch_size. I saw the method in [https://github.com/ultralytics/yolov3](url)

No, it won't help. BN's runing mean/std is calculated on forward, not backward. As long as you forward a small batch, it won't make any difference.

**Comment by DUZHUJUN** [comment]
@zylo117 change to a bigger batch size solved the problem,Thanks for your support.

**Comment by zylo117** [comment]
sorry for the training troubles, there's a bug in loss function. please pull the latest code and give it a try.
